Выборы в госдуму, 7% барьер

Maksonchik88 opened the pull request:
https://informatics.msk.ru/mod/statements/view.php?id=5763&chapterid=111354#1

Maksonchik88 committed:
commit_1

Maksonchik88 committed:
commit_2

iliadmitriev commented on files/seven_percent_barrie.py:
```diff
@@ -0,0 +1,22 @@
+
+parties_and_votes = dict(zip(parties_list, count_voces_list))
+
+total = sum(list(parties_and_votes.values()))
```
тут не надо list, потому что `sum` на вход может принимать не только список но и генератор

iliadmitriev commented on files/seven_percent_barrie.py:
```diff
@@ -0,0 +1,22 @@
+parties_list = lines[lines.index("PARTIES:") + 1: lines.index("VOTES:")]
+votes_list = lines[lines.index("VOTES:") + 1:]
+
+count_voces_list = []
+count_voces_list.append(votes_list.count('Party one'))
+count_voces_list.append(votes_list.count('Party two'))
+count_voces_list.append(votes_list.count('Party three'))
+
+parties_and_votes = dict(zip(parties_list, count_voces_list))
```
Решено не верно.
1. Если во входном файле будет 4 партии? Или партия поменяет название?
2. Ты используешь функцию zip для того чтобы склеить названия партий с количеством голосов, но ведь они могу быть расположены в совершенно произвольном порядке.

iliadmitriev commented on files/seven_percent_barrie.py:
```diff
@@ -0,0 +1,22 @@
+
+file.close()
+
+lines = list(filter(None, (line.rstrip() for line in lines)))
```
идея хорошая.
а как работает filter? и почему в параметрах `None`?

Maksonchik88 replied:
Идея заключается в том, что, применяя filter(), мы не указываем параметр function. Таким образом мы фильтруем весь текст файла, применяя строчный метод rstrip() и фильтруем каждую строку, сохраняя ее последовательно в список

Maksonchik88 replied:
а filter() на деле возвращает True-значения последовательности. В нашей задаче это все строки с применением метода rstrip().

Maksonchik88 replied:
Разобрался)))

iliadmitriev replied:
что значит True-значения? он что ли возвращает boolean? у нас же вроде как строки?

Maksonchik88 replied:
мы в функцию ничего не инициализировали, поэтому там None. при этом мы итерировали каждый элемент последовательности

Maksonchik88 replied:
нет, он не возвращает булевое значение, конечно (хотя, может очевидно)). в нашем случае на filter() всё True, поэтому все без исключения элементы последоватльности будут поочередно проитеррированы

iliadmitriev replied:
тоесть, он не нужен получается ?

Maksonchik88 replied:
ну конечно можно обычным циклом так сделать. По сути, метод можно не использовать. Но выглядит красиво.

iliadmitriev replied:
> все без исключения элементы последоватльности будут поочередно проитеррированы

Если ты используешь все элементы последовательности и ничего не фильтруешь, то зачем в принципе фильтровать ? Зачем эта строчка в коде, даже если она "выглядит красиво"?

iliadmitriev reviewed (changes requested):
Решено не верно.
Поправь заголовок pull request'а чтобы было понятно.

Maksonchik88 committed:
*Инициализировал словарь, в котором ключами являются партии, а значениями нули
*Проитеррировал список votes_list и заполнио созданный словарь количеством голосов для каждой партии
*Так же в данном цикле инициализировал и посчитал глобальную переменную - общее количество голосов

iliadmitriev commented on files/seven_percent_barrie.py:
```diff
@@ -0,0 +1,20 @@
+parties_list = lines[lines.index("PARTIES:") + 1: lines.index("VOTES:")]
+votes_list = lines[lines.index("VOTES:") + 1:]
+
+data_parties_keys = dict.fromkeys(parties_list, 0)
+
+global_votes = 0
+for voice in votes_list:
+    global_votes += 1
+    data_parties_keys[voice] += 1
```
Вот это сделал правильно и оно будет работать. 
Но не учел один кейс: 
Что если в списке голосов появится голос за несуществующую партию (испорченный билютень так сказать)
То у тебя автоматом этот голос добавится в качестве ключа к партиям и еще этот голос учтется в глобальных голосах

Maksonchik88 committed:
* Убрал filter()
* Добавил в цикл, где формируется глобальная переменная условие, которое исключит ошибку в случае, если будет "испорченная бюллетень")

iliadmitriev commented on files/seven_percent_barrie.py:
```diff
@@ -0,0 +1,23 @@
+    if voice in data_parties_keys:
+        global_votes += 1
+        data_parties_keys[voice] += 1
+    else:
+        continue
```
А это зачем?

Maksonchik88 replied:
Хз_))

iliadmitriev replied:
На работу кода это не влияет, но такое лучше не оставлять

iliadmitriev reviewed (approved):
Принято с мелким замечанием.
